### PR TITLE
Extend thinking time if max Q is not max N.

### PR DESCRIPTION
**Pull request by mooskagh**
(also possible to add other conditions to extend the search).

So far in my sample of 2 games it didn't perform very well.

new params to tune:
`--time-manager=smooth-experimental(init-piggybank=0.1,per-move-piggybank=0.05,max-piggybank-use=0.7,max-piggybank-moves=10.0)`

### CODE REVIEW

File: src/mcts/search.cc [hunk 1]
```diff
@@ -754,9 +754,30 @@ void Search::PopulateCommonIterationStats(IterationStats* stats) {
...
+
+  const auto draw_score = GetDrawScore(true);
+  const float fpu =
+      GetFpu(params_, root_node_, /* is_root_node */ true, draw_score);
+  float max_q = -1000;
+  uint64_t max_n = 0;
+  bool max_n_has_max_q = true;
   for (const auto& edge : root_node_->Edges()) {
-    stats->edge_n.push_back(edge.GetN());
+    const auto n = edge.GetN();
```
Comment (Tilps) [inline src/mcts/search.cc:766]: Wouldn't it make sense to make this max Q+M_effect isn't max N instead of just max Q ?

Comment (mooskagh) [reply src/mcts/search.cc:815]: Done.

**Comment by AlexisOlson** [comment]
**v25:** lc0 v0.25.1 -- full defaults except move overhead = 50ms
**PR1310:** Appveyor build for 673b99a -- full defaults except move overhead = 50ms
**Hardware:**  i7-6700K @ 4.0 GHz + RTX 2080
**Time control:** 5" + 0.05" (several thousand npm)
**Book:** All 960 positions, both sides
**Comment:** No detectable difference in Elo and time management was not discernable by watching. I definitely used the previous commit's binary but I'm not convinced it actually defaults to different behavior based on what the testing looks like.

```
   # PLAYER               :  RATING  ERROR   (%)  PLAYED  CFS(%)    W    L     D  D(%)
   1 lc0.v25.714428       :     0.0   ----    50    1920      53  317  315  1288    67
   2 lc0.PR1310.714428    :    -0.4    8.9    50    1920     ---  315  317  1288    67

White advantage = 30.75 +/- 4.57
Draw rate (equal opponents) = 67.89 % +/- 1.08
```

[PR1310_5s.zip](https://github.com/LeelaChessZero/lc0/files/4743142/PR1310_5s.zip)

**Comment by mooskagh** [comment]
> **Comment:** No detectable difference in Elo and time management was not discernable by watching. I definitely used the previous commit's binary but I'm not convinced it actually defaults to different behavior based on what the testing looks like.

Hm that's weird, that's not what other tests shown: https://discord.com/channels/425419482568196106/530486338236055583/718580264212758538

It's possible that either this PR broke normal time control, or something else that negatively affects strength was merged (either into master or in pr1310).

Could you test v0.25 vs master, or v0.25 vs pr1310 with old tc?.

**Comment by AlexisOlson** [comment]
It's going to be a while before I have RTX to spare again. Might try some tiny nets on other PC.

**Comment by lp200** [comment]
This PR is similar to the behavior of the original AlphaGo.

**Expansion (Figure 4b).
  It extends the search if the action maximizing visit count and the action maximizing action-value disagree.**

**Comment by AlexisOlson** [comment]
The test is finished. Here's what the time curves look like with "smooth" shown with dashed lines:

image

There's some averaging over multiple plies to smooth out the graphs. Lower is the average of times less than the overall average and upper is the average of times greater than the overall average.

**Match:** PR1310 vs 25.1 and master -- Chess960
**LC0 versions:** Full defaults for each binary except move overhead lowered to 50 ms.
**Hardware:**  i7-8700K @ 2.4 GHz + NVIDIA Quadro P620
**Time control:** 5" + 0.05"
**Book:** All Chess960 openings with reverse
**Tablebases:** 5-man syzygy
**Adjudication:** Syzygy Tablebase; Draw after 50 moves and 10 count under 10cp; Resign at 10 count over 1000cp
**Software:** Cutechess (GUI)
**Comment:** At least for short TC, this seems pretty solid.
**Ordo:** `ordo -Q -W -D -G -J -o summary.txt -j details.txt -s 10000 -U "0,1,2,5,4,6,7,9,8,10" -p "PR1310RR.pgn" -a 0 -A "lc0_25-1_69105b4e"`

```
   # PLAYER                 :  RATING  ERROR   (%)  PLAYED  CFS(%)     W     L     D  D(%)
   1 lc0_PR1310_67b99a      :    21.4   10.7    53    3840     100  1448  1204  1188    31
   2 lc0_25-1_69105b4e      :     0.0   ----    49    3840      64  1267  1373  1200    31
   3 lc0_master_07ddda19    :    -2.0   10.7    48    3840     ---  1234  1372  1234    32

White advantage = 12.88 +/- 3.79
Draw rate (equal opponents) = 31.53 % +/- 0.61
```

Head to head statistics:
```
1) lc0_PR1310_67b99a   21.4 :   3840 (+1448,=1188,-1204),  53.2 %

   vs.                        :  games (    +,    =,    -),   (%) :    Diff,    SD, CFS (%)
   lc0_25-1_69105b4e          :   1920 (  735,  577,  608),  53.3 :   +21.4,   5.5,  100.0
   lc0_master_07ddda19        :   1920 (  713,  611,  596),  53.0 :   +23.3,   5.4,  100.0

2) lc0_25-1_69105b4e    0.0 :   3840 (+1267,=1200,-1373),  48.6 %

   vs.                        :  games (    +,    =,    -),   (%) :    Diff,    SD, CFS (%)
   lc0_PR1310_67b99a          :   1920 (  608,  577,  735),  46.7 :   -21.4,   5.5,    0.0
   lc0_master_07ddda19        :   1920 (  659,  623,  638),  50.5 :    +2.0,   5.5,   63.9

3) lc0_master_07ddda19 -2.0 :   3840 (+1234,=1234,-1372),  48.2 %

   vs.                        :  games (    +,    =,    -),   (%) :    Diff,    SD, CFS (%)
   lc0_PR1310_67b99a          :   1920 (  596,  611,  713),  47.0 :   -23.3,   5.4,    0.0
   lc0_25-1_69105b4e          :   1920 (  638,  623,  659),  49.5 :    -2.0,   5.5,   36.1
```
[PR1310-RR.zip](https://github.com/LeelaChessZero/lc0/files/4755205/PR1310-RR.zip)

**Comment by Tilps** [comment]
As discussed offline - due to the evidence that the gains are only at short time control and not generalizing - the change should go back to default legacy.

File: src/mcts/search.cc [hunk 2]
```diff
@@ -70,6 +70,63 @@ MoveList MakeRootMoveFilter(const MoveList& searchmoves,
   return root_moves;
 }
 
+class MEvaluator {
```
Comment (Tilps) [inline src/mcts/search.cc:73]: I assume we're confident that the optimizer will optimize this away entirely....

File: src/mcts/search.cc [hunk 3]
```diff
@@ -70,6 +70,63 @@ MoveList MakeRootMoveFilter(const MoveList& searchmoves,
...
+
+  void SetParent(const Node* parent) {
+    assert(parent);
+    if (enabled_) {
+      parent_m_ = parent->GetM();
+      parent_within_threshold_ = WithinTheshold(parent, q_threshold_);
+    }
+  }
+
+  float GetM(const EdgeAndNode& child, float q) const {
```
Comment (Tilps) [inline src/mcts/search.cc:105]: I'd rather this be called GetMEffect or similar - GetMBoost - GetM has a clear existing meaning.

File: src/mcts/search.cc (hunk 2 above)
Comment (Tilps) [reply src/mcts/search.cc:73]: I'll follow up on this.

File: src/mcts/search.cc (hunk 3 above)
Comment (Tilps) [reply src/mcts/search.cc:105]: But I don't care that much...

**Comment by Tilps** [comment]
I changed the default back and merged master - will submit once it passes checks.

**Pull request merged**